libs/stm32f0_drivers/usart.c

```diff
@@ -0,0 +1,295 @@
+
+	/* Initialize tx/rx buffers */
+	u->buffers.tx = xQueueCreate(UART_BUFFER_SIZE, sizeof(uint8_t));
+	if (!u->buffers.tx)
+		return -ENOMEM;
+
+	u->buffers.rx = xQueueCreate(UART_BUFFER_SIZE, sizeof(uint8_t));
+	if (!u->buffers.rx)
+		return -ENOMEM;
+
+	u->gpios.rx_clock_cmd(u->gpios.rx_pin_clock, ENABLE);
+	u->gpios.tx_clock_cmd(u->gpios.tx_pin_clock, ENABLE);
+	u->rcc.periph_clock_cmd(u->rcc.rccperiph, ENABLE);
+
+	/* Enable the appropriate NVIC Channels */
+	nvic_init.NVIC_IRQChannel = u->irq_no;
+	nvic_init.NVIC_IRQChannelPriority = 5;
+	nvic_init.NVIC_IRQChannelCmd = ENABLE;
+
+	NVIC_Init(&nvic_init);
+
+	/* Set up the GPIOs */
+	gpio_conf.GPIO_Speed = GPIO_Speed_50MHz;
+	gpio_conf.GPIO_PuPd = GPIO_PuPd_UP;
+
+	/*Tx*/
+	gpio_conf.GPIO_Pin = u->gpios.txpin;
+	gpio_conf.GPIO_Mode = GPIO_Mode_AF;
+	gpio_conf.GPIO_OType = GPIO_OType_PP ;
+	GPIO_Init(u->gpios.txport, &gpio_conf);
+	GPIO_SetBits(u->gpios.txport, u->gpios.txpin);
+
+	/*Rx*/
+	gpio_conf.GPIO_Pin = u->gpios.rxpin;
+	gpio_conf.GPIO_Mode = GPIO_Mode_AF;
+	gpio_conf.GPIO_OType = GPIO_OType_PP ;
+	GPIO_Init(u->gpios.rxport, &gpio_conf);
+	GPIO_SetBits(u->gpios.rxport, u->gpios.txpin);
+
+	/* Configure GPIOS as AF */
+	GPIO_PinAFConfig(u->gpios.txport, u->gpios.tx_pinsource, GPIO_AF_1);
+	GPIO_PinAFConfig(u->gpios.rxport, u->gpios.rx_pinsource, GPIO_AF_1);
+
+	uart_conf.USART_BaudRate = baud;
+	uart_conf.USART_WordLength = USART_WordLength_8b;
+	uart_conf.USART_StopBits = USART_StopBits_1;
+	uart_conf.USART_Parity = USART_Parity_No;
+	uart_conf.USART_HardwareFlowControl = USART_HardwareFlowControl_None;
+	uart_conf.USART_Mode = USART_Mode_Rx | USART_Mode_Tx;
+
+	USART_Init(u->periph, &uart_conf);
+
+	USART_Cmd(u->periph, ENABLE);
+
+	USART_ITConfig(u->periph, USART_IT_RXNE, ENABLE);
+	u->initialized = true;
+
+	return 0;
+}
+
+int usart_enable_autocrlf(int usart_no, bool enabled)
+{
+	struct usart *u = usart_lookup(usart_no);
+
+	if (!u)
+		return -ENODEV;
+
+	u->autocrlf = enabled;
+
+	return 0;
+}
+
+int usart_write(int usart_no, char *s, int len)
+{
+	portBASE_TYPE q_ret;
+	int ret = 0;
+	static char linefeed = '\r';
+
+	struct usart *u = usart_lookup(usart_no);
+
+	if (!u)
+		return -ENODEV;
+
+	if(!u->initialized)
+		return -ENOTCONN;
+
+	while (len--) {
+		if (*s == '\n' && u->autocrlf) {
+			q_ret = xQueueSend(u->buffers.tx, &linefeed, portMAX_DELAY);
+			q_ret &= xQueueSend(u->buffers.tx, s, portMAX_DELAY);
+		} else {
+			q_ret = xQueueSend(u->buffers.tx, s, portMAX_DELAY);
+		}
+
+		s++;
+		ret++;
+
+		if (q_ret != pdTRUE) {
+			return -EIO;
+		} else {
+			USART_ITConfig(u->periph, USART_IT_TXE, ENABLE);
+		}
+	}
+
+	return ret;
+}
+
+int usart_putchar(int usart_no, char c)
+{
+	return usart_write(usart_no, &c, 1);
+}
+
+int usart_read(int usart_no, char *d, int len)
+{
+	int ret = 0;
+	struct usart *u = usart_lookup(usart_no);
+
+	if (!u)
+		return -ENODEV;
+
+	if(!u->initialized)
+		return -ENOTCONN;
+
+	while (len--) {
+		xQueueReceive(u->buffers.rx, d++, portMAX_DELAY);
```

Comment:
xQueueSend() can fail, hence why you check it with q_ret above. Can xQueueReceive() fail as well?